src/main/java/org/jenkinsci/plugins/KeycloakAccess.java

```diff
@@ -228,18 +229,77 @@ public UserDetails authenticate( String username, String password ) throws Authe
 	}
 
 	private String getAuthToken() {
-		String token = cache.getSystemToken();
-		if (token == null) {
-				Configuration configuration = new Configuration( keycloakDeployment.getAuthServerBaseUrl(), keycloakDeployment.getRealm(), keycloakDeployment.getResourceName(), keycloakDeployment.getResourceCredentials(), keycloakDeployment.getClient() );
-				AuthzClient authzClient = AuthzClient.create( configuration );
-				AccessTokenResponse accessTokenResponse = authzClient.obtainAccessToken();
-				token = accessTokenResponse.getToken();
-				LOGGER.finest( "Token expires in: " + accessTokenResponse.getExpiresIn() );
-				cache.setSystemToken( token, accessTokenResponse.getExpiresIn() );
+		String token = null;
+		SystemTokenInfo systemToken = cache.getSystemAccessToken();
+		if (systemToken != null) {
+			long now = System.currentTimeMillis();
+			if ( now > systemToken.getTokenExpirationTime() ) {
+				//Attempt to refresh
+				if ( now <= systemToken.getRefreshTokenExpirationTime() ) {
+					LOGGER.fine("Refreshing token");
+					AccessTokenResponse newToken = refreshToken( systemToken.getAccessToken().getRefreshToken() );

```

Comment:
Should this be surrounded by a try/catch to prevent exceptions from breaking the flow? It looks like the execution flow of this Http class throws an HttpResponseException on error response codes (i.e. anything <200 or >=300) - https://github.com/keycloak/keycloak/blob/master/authz/client/src/main/java/org/keycloak/authorization/client/util/HttpMethod.java#L95 as well as it can throw a generic runtime exception for other errors. The flow you've got here looks good and would be resilient enough to get through that if the exception isn't propagated back up.

Reply:
Added a try/catch to handle errors when retrieving the refresh token, such that the code will fall back to getting a new token.